packages/model-viewer/src/features/annotation.ts

```diff
@@ -0,0 +1,264 @@
+  }
+
+  /**
+   * Call this when adding elements to the same slot to keep track.
+   */
+  increment() {
+    ++this.referenceCount;
+  }
+
+  /**
+   * Call this when removing elements from the slot; returns true when the slot
+   * is unused.
+   */
+  decrement(): boolean {
+    return --this.referenceCount <= 0;
+  }
+
+  /**
+   * Change the position of the hotspot to the input string, in the same format
+   * as the data-position attribute.
+   */
+  updatePosition(position?: string) {
+    if (position == null)
+      return;
+    const positionNodes = parseExpressions(position)[0].terms;
+    for (let i = 0; i < 3; ++i) {
+      this.position.setComponent(
+          i, normalizeUnit(positionNodes[i] as NumberNode<'m'>).number);
+    }
+  }
+
+  /**
+   * Change the hotspot's normal to the input string, in the same format as the
+   * data-normal attribute.
+   */
+  updateNormal(normal?: string) {
+    if (normal == null)
+      return;
+    const normalNodes = parseExpressions(normal)[0].terms;
+    for (let i = 0; i < 3; ++i) {
+      this.normal.setComponent(
+          i, normalizeUnit(normalNodes[i] as NumberNode<'m'>).number);
+    }
+  }
+}
+
+export declare interface AnnotationInterface {
+  updateHotspot(config: HotspotConfiguration): void;
+}
+
+/**
+ * AnnotationMixin implements a declarative API to add hotspots and annotations.
+ * Child elements of the <model-viewer> element that have a slot name that
+ * begins with "hotspot" and data-position and data-normal attributes in
+ * the format of the camera-target attribute will be added to the scene and
+ * track the specified model coordinates.
+ */
+export const AnnotationMixin = <T extends Constructor<ModelViewerElementBase>>(
+    ModelViewerElement: T): Constructor<AnnotationInterface>&T => {
+  class AnnotationModelViewerElement extends ModelViewerElement {
+    private[$annotationRenderer] = new CSS2DRenderer();
+    private[$hotspotMap] = new Map<string, Hotspot>();
+    private[$mutationCallback] = (mutations: Array<unknown>) => {
+      mutations.forEach((mutation) => {
+        // NOTE: Be wary that in ShadyDOM cases, the MutationRecord
+        // only has addedNodes and removedNodes (and no other details).
+        if (!(mutation instanceof MutationRecord) ||
+            mutation.type === 'childList') {
+          (mutation as MutationRecord).addedNodes.forEach((node) => {
+            this[$addHotspot](node);
+          });
+          (mutation as MutationRecord).removedNodes.forEach((node) => {
+            this[$removeHotspot](node);
+          });
+        }
+      });
+    };
+    private[$observer] = new MutationObserver(this[$mutationCallback]);
+
+    constructor(...args: Array<any>) {
+      super(...args);
+
+      const {domElement} = this[$annotationRenderer];
+      domElement.classList.add('annotation-container');
+      this.shadowRoot!.querySelector('.container')!.appendChild(domElement);
+
+      for (let i = 0; i < this.children.length; ++i) {
```

Comment:
I think this loop needs to be done in the `connectedCallback`, otherwise there are conditions where the timing would cause some children not to be detected. Consider this:

```typescript
const mv = new ModelViewerElement(); // Loop runs to look for children, none found

for (let i = 0; i < numHotspots; ++i) {
  mv.appendChild(hotspotNode); // MutationObserver is not active yet
}

document.body.appendChild(mv); // Element is "connected"; MutationObserver is activated now
```